stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -243,6 +246,186 @@ gp_exp_quad_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
+
+  size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  for (size_t n = 0; n < x.size(); ++n)
+    check_not_nan("gp_exp_quad_cov", "x", x[n]);
+
+  matrix_cl x_gpu(x);
+  matrix_cl cov_gpu = gp_exp_quad_cov(x_gpu, sigma, length_scale);
+  copy(cov, cov_gpu);  // NOLINT
+
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of Eigen vectors of scalars.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale std::vector length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T = void>  // if this was non-templated overload or fully
+                              // specialized template, the compiler could
+                              // incorrectly resolve some called functions to
+                              // prim implementation instead of arr
+inline Eigen::MatrixXd gp_exp_quad_cov(
+    const std::vector<Eigen::VectorXd> &x, const double sigma,
+    const std::vector<double> &length_scale) {
+  check_positive_finite("gp_exp_quad_cov", "magnitude", sigma);
+  check_positive_finite("gp_exp_quad_cov", "length scale", length_scale);
+
+  size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  size_t l_size = length_scale.size();
+  check_size_match("gp_exp_quad_cov", "x dimension", x[0].size(),
+                   "number of length scales", l_size);
+
+  std::vector<Eigen::VectorXd> x_new = divide_columns(x, length_scale);
+
+  matrix_cl x_gpu(x_new);
+  matrix_cl cov_gpu = gp_exp_quad_cov(x_gpu, sigma, 1);
+  copy(cov, cov_gpu);  // NOLINT
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * This function is for the cross covariance matrix
+ * needed to compute posterior predictive density.
+ *
+ * @tparam T_x1 type of first std::vector of scalars
+ * @tparam T_x2 type of second std::vector of scalars
+ *    This function assumes each element of x1 and x2 are the same size.
+ * @tparam Cond condition
+ *
+ * @param x1 std::vector of elements that can be used in square distance
+ * @param x2 std::vector of elements that can be used in square distance
+ * @param sigma standard deviation
+ * @param length_scale length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T_x1, typename T_x2,
+          typename Cond = typename std::enable_if<
+              is_constant<T_x1>::value && is_constant<T_x2>::value>::type>
+inline typename Eigen::MatrixXd gp_exp_quad_cov(const std::vector<T_x1> &x1,
+                                                const std::vector<T_x2> &x2,
+                                                const double sigma,
+                                                const double length_scale) {
+  const char *function_name = "gp_exp_quad_cov";
```

Comment:
@syclik is the above alright or is using `__func__` going to be a goof on some compilers?

Reply:
That's right. I actually didn't know about the `__func__` keyword (as part of the C99) standard. But... to stay consistent with our codebase, we should just use `const char *function_name`. If you want this to change, create a new issue and we can try to make the change globally. I don't know how different compilers implement it, so it might not be ideal to use it. Or it might make our lives a ton easier. But we shouldn't just introduce it here in one function if we're not going to do it across our codebase.

Reply:
Word! @t4c1 sorry for the confusion can you switch this back?